File: mirea/cources/2026-2027/aii-july/01/Part-03-Sorting/examples/example-10.py
# Сортировка слиянием в Python3


def merge_sort(a):
    if len(a) <= 1:
        return a

    mid = len(a) // 2

    # Разбиваем список на две половины и сортируем их отдельно
    left = merge_sort(a[:mid])
    right = merge_sort(a[mid:])

    # Сливаем две уже отсортированные половины в один отсортированный список
    merged = []
    i = 0
    j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    # Добавляем оставшиеся элементы
    while i < len(left):
        merged.append(left[i])
        i += 1
    while j < len(right):
        merged.append(right[j])
        j += 1

    return merged


# Чтение списка и запуск сортировки
n = int(input())
a = list(map(int, input().split()))
a = merge_sort(a)

# Вывод отсортированного списка
print(*a)

File: mirea/cources/2026-2027/aii-july/01/Part-03-Sorting/examples/example-10.cpp
// Сортировка слиянием в C++

#include <iostream>
#include <vector>

using namespace std;

vector<int> merge_sort(const vector<int>& a) {
    if (a.size() <= 1) {
        return a;
    }

    int mid = (int)a.size() / 2;

    // Разбиваем массив на две половины и сортируем их отдельно
    vector<int> left(a.begin(), a.begin() + mid);
    vector<int> right(a.begin() + mid, a.end());
    left = merge_sort(left);
    right = merge_sort(right);

    // Сливаем две уже отсортированные половины в один отсортированный массив
    vector<int> merged;
    merged.reserve(a.size());

    int i = 0;
    int j = 0;
    while (i < (int)left.size() && j < (int)right.size()) {
        if (left[i] <= right[j]) {
            merged.push_back(left[i]);
            i++;
        } else {
            merged.push_back(right[j]);
            j++;
        }
    }

    // Добавляем оставшиеся элементы той половины, которая ещё не закончилась
    while (i < (int)left.size()) {
        merged.push_back(left[i]);
        i++;
    }
    while (j < (int)right.size()) {
        merged.push_back(right[j]);
        j++;
    }

    return merged;
}

int main() {
    // Чтение массива и запуск сортировки
    int n;
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    a = merge_sort(a);

    // Вывод отсортированного массива
    for (int x : a) {
        cout << x << ' ';
    }
    cout << '\n';
}